src/layout.rs

```diff
@@ -1,18 +1,88 @@
 use std::cmp::Ordering;
 
+/// Tell how grapheme clusters are supported / rendered.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum GraphemeClusterMode {
+    /// Support grapheme clustering
+    Unicode,
+    /// Doesn't support shaping
+    WcWidth,
+    /// Skip zero-width joiner
+    NoZwj,
+}
+
+impl GraphemeClusterMode {
+    /// Return default
+    #[cfg(test)]
+    pub fn from_env() -> Self {
+        GraphemeClusterMode::default()
+    }
+
+    /// Use environment variables to guess current mode
+    #[cfg(not(test))]
+    pub fn from_env() -> Self {
+        let gcm = match std::env::var("TERM_PROGRAM").as_deref() {
+            Ok("Apple_Terminal") => GraphemeClusterMode::Unicode,
+            Ok("iTerm.app") => GraphemeClusterMode::Unicode,
+            Ok("WezTerm") => GraphemeClusterMode::Unicode,
+            Err(std::env::VarError::NotPresent) => match std::env::var("TERM").as_deref() {
+                Ok("xterm-kitty") => GraphemeClusterMode::NoZwj,
```

Comment:
It is almost perfect. The one issue I had was with Kitty (on Debian):

![Image](https://github.com/user-attachments/assets/013bce24-5f58-4ae1-a753-7a6f5e364e83)

This seems to work better with:

```suggestion
                Ok("xterm-kitty") => GraphemeClusterMode::Unicode,
```

Reply:
Maybe the rendering varies depending on the platform / font used by Kitty ?
Because, I am almost sure that I tested kitty on MacOS. I will double check.
And, on Windows, all seems wrong:
```
Alacritty ~ (TERM=xterm-256color)
contour ~ (TERMINAL_NAME=contour)
WezTerm ~
Windows Terminal ~
```
But anyway, thanks for your feedback.

Reply:
On MacOS, kitty seems fine:
![image](https://github.com/user-attachments/assets/62b876d4-64e1-41f0-88b2-0685ff52c9c5)
(at least rustyline behaves like zsh)

Reply:
I tried changing the font for Kitty, but my understanding is that it just automatically use a fallback font for emojis if it is not present in the provided font.

I suspect that the issue I have observed does depend on the installed font, and I do not see how this could be handled 100% of the time. When testing in `xfce4-terminal`, My guess is that the underlying terminal emulator (GNOME’s [`vte`](https://github.com/GNOME/vte)) does something that prevent the combination of the separate emojis, regardless of the font. It is unclear whether that is intentional, but it does mean that the behavior stays consistent accross installations.

Anyway, having a bunch of empty space next to the character is not a terrible failure case. It is probably better than the original behavior where it was hard to know what you were even typing. So, probably a worthwile trade-off.